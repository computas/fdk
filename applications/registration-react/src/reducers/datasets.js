import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../constants/ActionTypes';

export default function datasets(
  state = { isFetchingDatasets: false, datasetItems: null },
  action
) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        isFetchingDatasets: true
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: action.response.data
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        isFetchingDatasets: false,
        datasetItems: null
      };
    default:
      return state;
  }
}
