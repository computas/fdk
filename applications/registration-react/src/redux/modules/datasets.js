import _ from 'lodash';
import {
  DATASETS_REQUEST,
  DATASETS_SUCCESS,
  DATASETS_FAILURE
} from '../../constants/ActionTypes';
import { fetchActions } from '../fetchActions';

function shouldFetch(metaState) {
  const threshold = 60 * 1000; // seconds
  return (
    !metaState ||
    (!metaState.isFetching &&
      (metaState.lastFetch || 0) < Date.now() - threshold)
  );
}

export function fetchDatasetsIfNeeded(catalogId) {
  return (dispatch, getState) =>
    shouldFetch(_.get(getState(), ['datasets', catalogId, 'meta'])) &&
    dispatch(
      fetchActions(`/catalogs/${catalogId}/datasets?size=100`, [
        { type: DATASETS_REQUEST, meta: { catalogId } },
        { type: DATASETS_SUCCESS, meta: { catalogId } },
        { type: DATASETS_FAILURE, meta: { catalogId } }
      ])
    );
}

const initialState = {};

export default function datasets(state = initialState, action) {
  switch (action.type) {
    case DATASETS_REQUEST:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: true,
            lastFetch: null
          }
        }
      };
    case DATASETS_SUCCESS:
      return {
        ...state,
        [action.meta.catalogId]: {
          items: _.get(action.payload, ['_embedded', 'datasets']),
          meta: {
            isFetching: false,
            lastFetch: Date.now()
          }
        }
      };
    case DATASETS_FAILURE:
      return {
        ...state,
        [action.meta.catalogId]: {
          meta: {
            isFetching: false,
            lastFetch: null
          }
        }
      };
    default:
      return state;
  }
}

export const getDatasetItemsByCatalogId = (datasets, catalogId) =>
  _.get(datasets, [catalogId, 'items']);
