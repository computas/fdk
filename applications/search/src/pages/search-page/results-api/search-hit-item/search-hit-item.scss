@import '../../../../../node_modules/designsystemet/fdk-designsystem-bootstrap4/scss/colors.scss';

.search-hit {
  background-color: $color0;
  border-radius: 5px;
  color: $color1;
  font-size: 1.5rem;
  padding: 1.5em 2em;
  margin-bottom: 1em;
  .search-hit__title-link {
    color: $color1;
    text-decoration: none;
  }
  .search-hit__version {
    background-color: $color-cta4;
    border-radius: 2px;
  }
}
