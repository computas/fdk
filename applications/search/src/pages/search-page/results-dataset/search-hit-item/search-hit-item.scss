
.fdk-a-search-hit {
  text-decoration: none;
}
