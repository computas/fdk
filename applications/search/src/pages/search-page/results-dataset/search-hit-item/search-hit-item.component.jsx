import React from 'react';
import PropTypes from 'prop-types';
import * as _ from 'lodash';
import cx from 'classnames';
import { Link } from 'react-router-dom';

import { DistributionFormat } from '../../../../components/distribution-format/distribution-format.component';
import localization from '../../../../lib/localization';
import { getTranslateText } from '../../../../lib/translateText';
import { getDistributionTypeByUri } from '../../../../redux/modules/distributionType';
import { SearchHitHeader } from '../../../../components/search-hit-header/search-hit-header.component';
import './search-hit-item.scss';

const renderFormats = (source, code, distributionTypeItems) => {
  const { distribution } = source;

  const children = (distributions, code) => {
    const nodes = [];
    distributions.forEach(item => {
      const { format } = item;
      let { type } = item;
      if (format && typeof format !== 'undefined') {
        if (
          type &&
          (type !== 'API' && type !== 'Feed' && type !== 'Nedlastbar fil')
        ) {
          const distributionType = getDistributionTypeByUri(
            distributionTypeItems,
            type
          );
          if (distributionType !== null && distributionType.length > 0) {
            type = getTranslateText(distributionType[0].prefLabel);
          } else {
            type = null;
          }
        }
        const formatNodes = Object.keys(format).map(key => (
          <DistributionFormat
            key={`dataset-distribution-format${key}`}
            code={code}
            text={format[key]}
            type={type}
          />
        ));
        nodes.push(formatNodes);
      }
    });
    return nodes;
  };

  if (distribution && _.isArray(Object.keys(distribution))) {
    return <div>{children(distribution, code)}</div>;
  }
  return null;
};

const renderSample = source => {
  const { sample } = source;
  if (sample) {
    if (sample.length > 0) {
      return <div id="search-hit-sample">{localization.search_hit.sample}</div>;
    }
  }
  return null;
};

export const SearchHitItem = props => {
  const { distributionTypeItems } = props;
  const { _source } = props.result;

  const hitId = encodeURIComponent(_source.id);
  const { publisher, theme, provenance } = _source;
  let { title, description, objective } = _source;
  if (title) {
    title = getTranslateText(_source.title);
  }
  if (description) {
    description = getTranslateText(_source.description);
  }
  if (objective) {
    objective = getTranslateText(_source.objective);
  }

  if (description && description.length > 220) {
    description = `${description.substr(0, 220)}...`;
  } else if (description && description.length < 150 && objective) {
    const freeLength = 200 - description.length;
    const objectiveLength = objective.length;
    description = `${description} ${objective.substr(0, 200 - freeLength)} ${
      objectiveLength > freeLength ? '...' : ''
    }`;
  }
  const link = `/datasets/${hitId}`;

  let accessRightsLabel;
  let distributionNonPublic = false;
  let distributionRestricted = false;
  let distributionPublic = false;

  let authorityCode = '';
  if (_source.accessRights && _source.accessRights.code) {
    authorityCode = _source.accessRights.code;
  }

  if (_source.accessRights && authorityCode === 'NON_PUBLIC') {
    distributionNonPublic = true;
    accessRightsLabel =
      localization.dataset.accessRights.authorityCode.nonPublic;
  } else if (_source.accessRights && authorityCode === 'RESTRICTED') {
    distributionRestricted = true;
    accessRightsLabel =
      localization.dataset.accessRights.authorityCode.restricted;
  } else if (_source.accessRights && authorityCode === 'PUBLIC') {
    distributionPublic = true;
    accessRightsLabel = localization.dataset.accessRights.authorityCode.public;
  }

  const distributionClass = cx({
    'fdk-container-distributions':
      distributionNonPublic || distributionRestricted || distributionPublic,
    'fdk-distributions-red': distributionNonPublic,
    'fdk-distributions-yellow': distributionRestricted,
    'fdk-distributions-green': distributionPublic
  });

  return (
    <article>
      <Link
        className="fdk-a-search-hit"
        title={`${localization.result.dataset}: ${title}`}
        to={link}
      >
        <span className="uu-invisible" aria-hidden="false">
          SÃ¸ketreff.
        </span>
        <div className="fdk-container-search-hit">
          <header>
            <SearchHitHeader
              tag="h2"
              title={title}
              publisherLabel={localization.search_hit.owned}
              publisher={publisher}
              theme={theme}
              provenance={provenance}
            />
          </header>
          <p className="fdk-p-search-hit">
            <span className="uu-invisible" aria-hidden="false">
              Beskrivelse av datasettet,
            </span>
            {description}
          </p>
          <div className={distributionClass}>
            <strong>{accessRightsLabel}</strong>
            {renderFormats(_source, authorityCode, distributionTypeItems)}
            {renderSample(_source)}
          </div>
        </div>
      </Link>
    </article>
  );
};

SearchHitItem.defaultProps = {
  result: null,
  distributionTypeItems: null
};

SearchHitItem.propTypes = {
  result: PropTypes.shape({}),
  distributionTypeItems: PropTypes.array
};
