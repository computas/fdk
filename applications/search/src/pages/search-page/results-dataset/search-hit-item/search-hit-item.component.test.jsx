import React from 'react';
import { shallow } from 'enzyme';
import { SearchHitItem } from './search-hit-item.component';
import datasets from '../../../../../test/fixtures/datasets';
import distributionTypes from '../../../../../test/fixtures/distributionTypes';

let defaultProps;

test('should render SearchHitItem NON_PUBLIC correctly', () => {
  defaultProps = {
    result: datasets[0]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem RESTRICTED correctly', () => {
  defaultProps = {
    result: datasets[1]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem PUBLIC correctly', () => {
  defaultProps = {
    result: datasets[2]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem with no title, description, accessRights and distribution with no format correctly', () => {
  defaultProps = {
    result: datasets[3]
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render SearchHitItem with distribution type correctly', () => {
  defaultProps = {
    result: datasets[5],
    distributionTypeItems: distributionTypes.distributionTypeItems
  };
  const wrapper = shallow(<SearchHitItem {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
