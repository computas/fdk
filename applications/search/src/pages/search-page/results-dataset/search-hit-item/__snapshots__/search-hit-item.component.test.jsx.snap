// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchHitItem NON_PUBLIC correctly 1`] = `
<article>
  <Link
    className="fdk-a-search-hit"
    replace={false}
    title="Dataset: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
    to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Søketreff.
    </span>
    <div
      className="fdk-container-search-hit"
    >
      <header>
        <SearchHitHeader
          provenance={
            Object {
              "code": "vedtak",
              "prefLabel": Object {
                "nb": "Vedtak",
              },
              "uri": "http://data.brreg.no/datakatalog/provenance/vedtak",
            }
          }
          publisher={
            Object {
              "id": "987654321",
              "name": "TESTETATEN",
              "uri": "http://data.brreg.no/enhetsregisteret/enhet/987654321",
            }
          }
          publisherItems={null}
          publisherLabel="Owned by"
          tag="h2"
          theme={
            Array [
              Object {
                "title": Object {
                  "nb": "Forvaltning og offentlig støtte",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/GOVE",
              },
              Object {
                "title": Object {
                  "nb": "Miljø",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/ENVI",
              },
            ]
          }
          title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
          titleLink={null}
        />
      </header>
      <p
        className="fdk-p-search-hit"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Beskrivelse av datasettet,
        </span>
        Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4...
      </p>
      <div
        className="fdk-container-distributions fdk-distributions-red"
      >
        <strong>
          Non public
        </strong>
        <div />
        <div
          id="search-hit-sample"
        >
          Sample data are available
        </div>
      </div>
    </div>
  </Link>
</article>
`;

exports[`should render SearchHitItem PUBLIC correctly 1`] = `
<article>
  <Link
    className="fdk-a-search-hit"
    replace={false}
    title="Dataset: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
    to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Søketreff.
    </span>
    <div
      className="fdk-container-search-hit"
    >
      <header>
        <SearchHitHeader
          provenance={
            Object {
              "code": "vedtak",
              "prefLabel": Object {
                "nb": "Vedtak",
              },
              "uri": "http://data.brreg.no/datakatalog/provenance/vedtak",
            }
          }
          publisher={null}
          publisherItems={null}
          publisherLabel="Owned by"
          tag="h2"
          theme={
            Array [
              Object {
                "title": Object {
                  "nb": "Forvaltning og offentlig støtte",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/GOVE",
              },
              Object {
                "title": Object {
                  "nb": "Miljø",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/ENVI",
              },
            ]
          }
          title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
          titleLink={null}
        />
      </header>
      <p
        className="fdk-p-search-hit"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Beskrivelse av datasettet,
        </span>
        Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commo ...
      </p>
      <div
        className="fdk-container-distributions fdk-distributions-green"
      >
        <strong>
          Public
        </strong>
      </div>
    </div>
  </Link>
</article>
`;

exports[`should render SearchHitItem RESTRICTED correctly 1`] = `
<article>
  <Link
    className="fdk-a-search-hit"
    replace={false}
    title="Dataset: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
    to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Søketreff.
    </span>
    <div
      className="fdk-container-search-hit"
    >
      <header>
        <SearchHitHeader
          provenance={
            Object {
              "code": "vedtak",
              "prefLabel": Object {
                "nb": "Vedtak",
              },
              "uri": "http://data.brreg.no/datakatalog/provenance/vedtak",
            }
          }
          publisher={
            Object {
              "id": "987654321",
              "name": "TESTETATEN",
              "uri": "http://data.brreg.no/enhetsregisteret/enhet/987654321",
            }
          }
          publisherItems={null}
          publisherLabel="Owned by"
          tag="h2"
          theme={
            Array [
              Object {
                "title": Object {
                  "nb": "Forvaltning og offentlig støtte",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/GOVE",
              },
              Object {
                "title": Object {
                  "nb": "Miljø",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/ENVI",
              },
            ]
          }
          title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
          titleLink={null}
        />
      </header>
      <p
        className="fdk-p-search-hit"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Beskrivelse av datasettet,
        </span>
        Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder i Oslo og nærliggende kommuner (markaloven) som trådte i kraft 1. september 2009. Markalovens virkeområde er fastsatt i forskrift 4...
      </p>
      <div
        className="fdk-container-distributions fdk-distributions-yellow"
      >
        <strong>
          Restricted
        </strong>
        <div>
          <DistributionFormat
            code="RESTRICTED"
            key="dataset-distribution-format0"
            text="application/json"
            type={null}
          />
          <DistributionFormat
            code="RESTRICTED"
            key="dataset-distribution-format1"
            text="xml"
            type={null}
          />
        </div>
        <div
          id="search-hit-sample"
        >
          Sample data are available
        </div>
      </div>
    </div>
  </Link>
</article>
`;

exports[`should render SearchHitItem with distribution type correctly 1`] = `
<article>
  <Link
    className="fdk-a-search-hit"
    replace={false}
    title="Dataset: AAMarkagrensen Oslo Kommune og nærliggende kommuner"
    to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Søketreff.
    </span>
    <div
      className="fdk-container-search-hit"
    >
      <header>
        <SearchHitHeader
          provenance={
            Object {
              "code": "vedtak",
              "prefLabel": Object {
                "nb": "Vedtak",
              },
              "uri": "http://data.brreg.no/datakatalog/provenance/vedtak",
            }
          }
          publisher={null}
          publisherItems={null}
          publisherLabel="Owned by"
          tag="h2"
          theme={
            Array [
              Object {
                "title": Object {
                  "nb": "Forvaltning og offentlig støtte",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/GOVE",
              },
              Object {
                "title": Object {
                  "nb": "Miljø",
                },
                "uri": "http://publications.europa.eu/resource/authority/data-theme/ENVI",
              },
            ]
          }
          title="AAMarkagrensen Oslo Kommune og nærliggende kommuner"
          titleLink={null}
        />
      </header>
      <p
        className="fdk-p-search-hit"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Beskrivelse av datasettet,
        </span>
        Datasettet avgrenser område for virkeområdet til lov 6. juni 2009 nr. 35 om naturområder Datasettes formål er nullam quis rius eget urna mollis ornare vel eu leo. Praesent commo ...
      </p>
      <div
        className="fdk-container-distributions fdk-distributions-green"
      >
        <strong>
          Public
        </strong>
        <div>
          <DistributionFormat
            code="PUBLIC"
            key="dataset-distribution-format0"
            text="xml"
            type="Downloadable file"
          />
          <DistributionFormat
            code="PUBLIC"
            key="dataset-distribution-format1"
            text="tekst"
            type="Downloadable file"
          />
          <DistributionFormat
            code="PUBLIC"
            key="dataset-distribution-format2"
            text="json"
            type="Downloadable file"
          />
          <DistributionFormat
            code="PUBLIC"
            key="dataset-distribution-format0"
            text="xml"
            type={null}
          />
          <DistributionFormat
            code="PUBLIC"
            key="dataset-distribution-format1"
            text="tekst"
            type={null}
          />
          <DistributionFormat
            code="PUBLIC"
            key="dataset-distribution-format2"
            text="json"
            type={null}
          />
        </div>
      </div>
    </div>
  </Link>
</article>
`;

exports[`should render SearchHitItem with no title, description, accessRights and distribution with no format correctly 1`] = `
<article>
  <Link
    className="fdk-a-search-hit"
    replace={false}
    title="Dataset: undefined"
    to="/datasets/3645d5a8-6689-4fc1-8684-d3d6c2d47386"
  >
    <span
      aria-hidden="false"
      className="uu-invisible"
    >
      Søketreff.
    </span>
    <div
      className="fdk-container-search-hit"
    >
      <header>
        <SearchHitHeader
          provenance={
            Object {
              "code": "vedtak",
              "prefLabel": Object {
                "nb": "Vedtak",
              },
              "uri": "http://data.brreg.no/datakatalog/provenance/vedtak",
            }
          }
          publisher={
            Object {
              "id": "987654321",
              "name": "TESTETATEN",
              "uri": "http://data.brreg.no/enhetsregisteret/enhet/987654321",
            }
          }
          publisherItems={null}
          publisherLabel="Owned by"
          tag="h2"
          theme={null}
          title={null}
          titleLink={null}
        />
      </header>
      <p
        className="fdk-p-search-hit"
      >
        <span
          aria-hidden="false"
          className="uu-invisible"
        >
          Beskrivelse av datasettet,
        </span>
      </p>
      <div
        className=""
      >
        <strong />
        <div />
      </div>
    </div>
  </Link>
</article>
`;
